modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.optimization;
+
+import org.apache.ignite.ml.math.Matrix;
+import org.apache.ignite.ml.math.Vector;
+import org.apache.ignite.ml.math.impls.matrix.SparseDistributedMatrix;
+import org.apache.ignite.ml.math.impls.vector.DenseLocalOnHeapVector;
+import org.apache.ignite.ml.math.impls.vector.FunctionVector;
+import org.apache.ignite.ml.optimization.util.SparseDistributedMatrixMapReducer;
+
+/**
+ * Gradient descent optimizer.
+ */
+public class GradientDescent {
+    /**
+     * Function which computes gradient of the loss function at any given point.
+     */
+    private final GradientFunction lossGradient;
+
+    /**
+     * Weights updater applied on every gradient descent step to decide how weights should be changed.
+     */
+    private final Updater updater;
+
+    /**
+     * Max number of gradient descent iterations.
+     */
+    private int maxIterations = 1000;
+
+    /**
+     * Convergence tolerance is condition which decides iteration termination.
+     */
+    private double convergenceTol = 1e-8;
+
+    /**
+     * New gradient descent instance based of loss function and updater.
+     *
+     * @param lossGradient Function which computes gradient of the loss function at any given point
+     * @param updater Weights updater applied on every gradient descent step to decide how weights should be changed
+     */
+    public GradientDescent(GradientFunction lossGradient, Updater updater) {
+        this.lossGradient = lossGradient;
+        this.updater = updater;
+    }
+
+    /**
+     * Sets max number of gradient descent iterations.
+     *
+     * @param maxIterations Max number of gradient descent iterations
+     * @return This gradient descent instance
+     */
+    public GradientDescent withMaxIterations(int maxIterations) {
+        if (maxIterations < 0)
```

Comment:
Mostly we use asserts to check input arguments, Let's discuss it with @ybabak @YuriBabak

Reply:
Replaced with asserts.